identity/aziot-identityd/src/identity.rs

```diff
@@ -830,3 +935,149 @@ impl HubDeviceInfo {
         "unprovisioned".to_owned()
     }
 }
+
+pub struct ModuleBackup {}
+
+impl ModuleBackup {
+    pub fn set_device(homedir_path: &Path, iothub_hostname: &str, device_id: &str) {
+        let result = Self::get_device_path(homedir_path, iothub_hostname, device_id)
+            .map_err(|err| Error::Internal(InternalError::GetModulePath(err)))
+            .and_then(|path| {
+                if !path.exists() {
+                    // Clean old device module backups
+                    let mut old_modules_path = homedir_path.to_owned();
+                    old_modules_path.push(MODULE_BACKUP_LOCATION);
+
+                    // Best effort to remove old modules backup folder, in case permissions have been set
+                    // explictly on filesystem to prevent removal
+                    if let Err(err) = std::fs::remove_dir_all(old_modules_path) {
+                        if err.kind() != std::io::ErrorKind::NotFound {
+                            log::warn!("Failed to clear old module backup state: {}", err);
+                        }
+                    }
+
+                    // Create new device's module backup folder
+                    return std::fs::create_dir_all(&path)
+                        .map_err(|err| Error::Internal(InternalError::SaveModuleBackup(err)));
+                }
+
+                Ok(())
+            });
+
+        // Logging a warning is sufficient for per-module backup to keep the service operational for online operations
+        if let Err(err) = result {
+            log::warn!(
+                "Failed to create module information backup state folder: {}",
+                err
+            );
+        }
+    }
+
+    pub fn set_module_backup(
+        homedir_path: &Path,
+        iothub_hostname: &str,
+        device_id: &str,
+        module_id: &str,
+        data: &Option<aziot_identity_common::hub::Module>,
```

Comment:
`&Option<T>` -> `Option<&T>`

Reply:
Turns out I didn't need to send a reference. I fixed some instances of this in `lib.rs` for `issue_local_identity` though.